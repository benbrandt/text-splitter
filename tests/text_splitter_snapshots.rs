use std::fs;

use more_asserts::assert_le;
use once_cell::sync::Lazy;
use text_splitter::{Characters, ChunkSizer, TextSplitter};
use tiktoken_rs::{cl100k_base, CoreBPE};
use tokenizers::Tokenizer;

#[test]
fn characters_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::default();
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(Characters.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn characters_range() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        let splitter = TextSplitter::default();
        let chunks = splitter.chunks(&text, 500..=2000).collect::<Vec<_>>();

        assert_eq!(chunks.join(""), text);
        for chunk in chunks.iter() {
            assert_le!(Characters.chunk_size(chunk), 2000);
        }
        insta::assert_yaml_snapshot!(chunks);
    });
}

#[test]
fn characters_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::default().with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

static HUGGINGFACE_TOKENIZER: Lazy<Tokenizer> =
    Lazy::new(|| Tokenizer::from_pretrained("bert-base-cased", None).unwrap());

#[test]
fn huggingface_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(HUGGINGFACE_TOKENIZER.clone());
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(HUGGINGFACE_TOKENIZER.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn huggingface_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(HUGGINGFACE_TOKENIZER.clone()).with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

static TIKTOKEN_TOKENIZER: Lazy<CoreBPE> = Lazy::new(|| cl100k_base().unwrap());

#[test]
fn tiktoken_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(TIKTOKEN_TOKENIZER.clone());
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(TIKTOKEN_TOKENIZER.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn tiktoken_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(TIKTOKEN_TOKENIZER.clone()).with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}
