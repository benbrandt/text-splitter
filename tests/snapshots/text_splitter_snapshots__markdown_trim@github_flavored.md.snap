---
source: tests/text_splitter_snapshots.rs
expression: chunks
input_file: tests/inputs/markdown/github_flavored.md
---
- "# Headers"
- "```"
- "# h1"
- Heading 8-
- )
- "## h2"
- Heading
- "### h3"
- Heading
- "#### h4"
- Heading
- "##### h5"
- Heading
- "###### h6"
- Heading
- Alternativ
- "ely, for"
- "H1 and H2,"
- an
- underline-
- "ish style:"
- Alt-H1
- "======"
- Alt-H2
- "------\n```"
- "# h1"
- Heading 8-
- )
- "## h2"
- Heading
- "### h3"
- Heading
- "#### h4"
- Heading
- "##### h5"
- Heading
- "###### h6"
- Heading
- Alternativ
- "ely, for"
- "H1 and H2,"
- an
- underline-
- "ish style:"
- Alt-H1
- "======"
- Alt-H2
- "------"
- "------"
- "# Emphasis"
- "```"
- "Emphasis,"
- aka
- "italics,"
- with *
- asterisks*
- or
- _underscor
- es_.
- Strong
- "emphasis,"
- "aka bold,"
- with **
- asterisks*
- "* or"
- __undersco
- res__.
- Combined
- emphasis
- with **
- asterisks
- and
- _underscor
- es_**.
- Strikethro
- ugh uses
- two tildes
- "."
- ~~Scratch
- this.~~
- "**This is"
- bold text*
- "*"
- __This is
- bold
- text__
- "*This is"
- italic
- text*
- _This is
- italic
- text_
- ~~
- Strikethro
- "ugh~~\n```"
- "Emphasis,"
- aka
- "italics,"
- with *
- asterisks*
- or
- _underscor
- es_.
- Strong
- "emphasis,"
- "aka bold,"
- with **
- asterisks*
- "* or"
- __undersco
- res__.
- Combined
- emphasis
- with **
- asterisks
- and
- _underscor
- es_**.
- Strikethro
- ugh uses
- two tildes
- "."
- ~~Scratch
- this.~~
- "**This is"
- bold text*
- "*"
- __This is
- bold
- text__
- "*This is"
- italic
- text*
- _This is
- italic
- text_
- ~~
- Strikethro
- ugh~~
- "------"
- "# Lists"
- "```\n1."
- First
- ordered
- list item
- "2."
- Another
- item
- ⋅⋅*
- Unordered
- sub-list.
- "1."
- Actual
- numbers
- "don't"
- "matter,"
- just that
- "it's a"
- number
- ⋅⋅1.
- Ordered
- sub-list
- "4."
- And
- another
- item.
- ⋅⋅⋅You can
- have
- properly
- indented
- paragraphs
- within
- list items
- "."
- Notice the
- blank line
- "above, and"
- the
- leading
- spaces (at
- "least one,"
- "but we'll"
- use three
- here to
- also align
- the raw
- Markdown).
- ⋅⋅⋅To have
- a line
- break
- without a
- "paragraph,"
- you will
- need to
- use two
- trailing
- spaces.⋅⋅
- ⋅⋅⋅Note
- that this
- line is
- "separate,"
- but within
- the same
- paragraph.
- ⋅⋅
- ⋅⋅⋅(This
- is
- contrary
- to the
- typical
- GFM line
- break
- "behaviour,"
- where
- trailing
- spaces are
- not
- required.)
- "*"
- Unordered
- list can
- use
- asterisks
- "- Or"
- minuses
- + Or
- "pluses\n\n1."
- Make my
- changes
- 1. Fix bug
- "2."
- Improve
- formatting
- "- Make the"
- headings
- "bigger\n2."
- Push my
- commits to
- "GitHub\n3."
- Open a
- pull
- request
- "* Describe"
- my changes
- "* Mention"
- all the
- members of
- my team
- "* Ask for"
- feedback
- + Create a
- list by
- starting a
- line with
- "`+`, `-`,"
- "or `*`"
- + Sub-
- lists are
- made by
- indenting
- "2 spaces:"
- "- Marker"
- character
- change
- forces new
- list start
- ":"
- "* Ac"
- tristique
- libero
- volutpat
- at
- +
- Facilisis
- in pretium
- nisl
- aliquet
- "- Nulla"
- volutpat
- aliquam
- velit
- + Very
- "easy!\n```"
- "1."
- First
- ordered
- list item
- "2."
- Another
- "item\n⋅⋅*"
- Unordered
- sub-list.
- "1."
- Actual
- numbers
- "don't"
- "matter,"
- just that
- "it's a"
- number
- ⋅⋅1.
- Ordered
- sub-list
- "4."
- And
- another
- item.
- ⋅⋅⋅You can
- have
- properly
- indented
- paragraphs
- within
- list items
- "."
- Notice the
- blank line
- "above, and"
- the
- leading
- spaces (at
- "least one,"
- "but we'll"
- use three
- here to
- also align
- the raw
- Markdown).
- ⋅⋅⋅To have
- a line
- break
- without a
- "paragraph,"
- you will
- need to
- use two
- trailing
- spaces.⋅⋅
- ⋅⋅⋅Note
- that this
- line is
- "separate,"
- but within
- the same
- paragraph.
- ⋅⋅
- ⋅⋅⋅(This
- is
- contrary
- to the
- typical
- GFM line
- break
- "behaviour,"
- where
- trailing
- spaces are
- not
- required.)
- "*"
- Unordered
- list can
- use
- asterisks
- "- Or"
- minuses
- + Or
- pluses
- "1."
- Make my
- changes
- 1. Fix bug
- "2."
- Improve
- formatting
- "- Make the"
- headings
- bigger
- "2."
- Push my
- commits to
- GitHub
- "3."
- Open a
- pull
- request
- "* Describe"
- my changes
- "* Mention"
- all the
- members of
- my team
- "* Ask for"
- feedback
- + Create a
- list by
- starting a
- line with
- "`+`, `-`,"
- "or `*`"
- + Sub-
- lists are
- made by
- indenting
- "2 spaces:"
- "- Marker"
- character
- change
- forces new
- list start
- ":"
- "* Ac"
- tristique
- libero
- volutpat
- at
- +
- Facilisis
- in pretium
- nisl
- aliquet
- "- Nulla"
- volutpat
- aliquam
- velit
- + Very
- easy!
- "------"
- "# Task"
- lists
- "```"
- "- [x]"
- Finish my
- changes
- "- [ ] Push"
- my commits
- to GitHub
- "- [ ] Open"
- a pull
- request
- "- [x] @"
- "mentions,"
- "#refs, ["
- "links](),"
- "**"
- formatting
- "**, and <"
- del>tags</
- del>
- supported
- "- [x] list"
- syntax
- required (
- any
- unordered
- or ordered
- list
- supported)
- "- [x] this"
- is a
- complete
- item
- "- [ ] this"
- is an
- incomplete
- "item\n```"
- "- [x]"
- Finish my
- changes
- "- [ ] Push"
- my commits
- to GitHub
- "- [ ] Open"
- a pull
- request
- "- [x] @"
- "mentions,"
- "#refs, ["
- "links](),"
- "**"
- formatting
- "**, and <"
- del>tags</
- del>
- supported
- "- [x] list"
- syntax
- required (
- any
- unordered
- or ordered
- list
- supported)
- "- [ ] this"
- is a
- complete
- item
- "- [ ] this"
- is an
- incomplete
- item
- "------"
- "# Ignoring"
- Markdown
- formatting
- You can
- tell
- GitHub to
- ignore (or
- escape)
- Markdown
- formatting
- "by using \\"
- before the
- Markdown
- character.
- "```"
- "Let's"
- "rename \\*"
- our-new-
- "project\\*"
- "to \\*our-"
- old-
- "project\\*."
- "```"
- "Let's"
- "rename \\*"
- our-new-
- "project\\*"
- "to \\*our-"
- old-
- "project\\*."
- "------"
- "# Links"
- "```"
- "[I'm an"
- inline-
- style link
- "](https://"
- www.google
- ".com)"
- "[I'm an"
- inline-
- style link
- with title
- "](https://"
- www.google
- ".com \""
- "Google's"
- "Homepage\")"
- "[I'm a"
- reference-
- style link
- "]["
- Arbitrary
- case-
- insensitiv
- e
- reference
- "text]"
- "[I'm a"
- relative
- reference
- to a
- repository
- "file](../"
- blob/
- master/
- LICENSE)
- "[You can"
- use
- numbers
- for
- reference-
- style link
- definition
- "s][1]"
- Or leave
- it empty
- and use
- "the [link"
- text
- "itself]."
- URLs and
- URLs in
- angle
- brackets
- will
- automatica
- lly get
- turned
- into links
- "."
- "http://"
- www.exampl
- e.com or <
- "http://"
- www.exampl
- e.com> and
- sometimes
- example.co
- m (but not
- "on Github,"
- for
- example).
- Some text
- to show
- that the
- reference
- links can
- follow
- later.
- "[arbitrary"
- case-
- insensitiv
- e
- reference
- "text]:"
- "https://"
- www.mozill
- a.org
- "[1]: http:"
- //
- slashdot.o
- rg
- "[link text"
- "itself]:"
- "http://"
- www.reddit
- ".com\n```"
- "[I'm an"
- inline-
- style link
- "](https://"
- www.google
- ".com)"
- "[I'm an"
- inline-
- style link
- with title
- "](https://"
- www.google
- ".com \""
- "Google's"
- "Homepage\")"
- "[I'm a"
- reference-
- style link
- "]["
- Arbitrary
- case-
- insensitiv
- e
- reference
- "text]"
- "[I'm a"
- relative
- reference
- to a
- repository
- "file](../"
- blob/
- master/
- LICENSE)
- "[You can"
- use
- numbers
- for
- reference-
- style link
- definition
- "s][1]"
- Or leave
- it empty
- and use
- "the [link"
- text
- "itself]."
- URLs and
- URLs in
- angle
- brackets
- will
- automatica
- lly get
- turned
- into links
- "."
- "http://"
- www.exampl
- e.com or <
- "http://"
- www.exampl
- e.com> and
- sometimes
- example.co
- m (but not
- "on Github,"
- for
- example).
- Some text
- to show
- that the
- reference
- links can
- follow
- later.
- "[arbitrary"
- case-
- insensitiv
- e
- reference
- "text]:"
- "https://"
- www.mozill
- a.org
- "[1]: http:"
- //
- slashdot.o
- rg
- "[link text"
- "itself]:"
- "http://"
- www.reddit
- ".com"
- "------"
- "# Images"
- "```"
- "Here's our"
- logo (
- hover to
- see the
- title text
- "):"
- Inline-
- "style:\n!["
- "alt text]("
- "https://"
- github.com
- /adam-p/
- markdown-
- here/raw/
- master/src
- /common/
- images/
- icon48.png
- "\"Logo"
- Title Text
- "1\")"
- Reference-
- "style:\n!["
- "alt text]["
- "logo]"
- "[logo]:"
- "https://"
- github.com
- /adam-p/
- markdown-
- here/raw/
- master/src
- /common/
- images/
- icon48.png
- "\"Logo"
- Title Text
- "2\"\n\n!["
- "Minion]("
- "https://"
- octodex.gi
- thub.com/
- images/
- minion.png
- ")\n!["
- Stormtroop
- "ocat]("
- "https://"
- octodex.gi
- thub.com/
- images/
- stormtroop
- "ocat.jpg \""
- The
- Stormtroop
- "ocat\")"
- Like links
- ", Images"
- also have
- a footnote
- style
- "syntax\n\n!["
- "Alt text]["
- "id]"
- With a
- reference
- later in
- the
- document
- defining
- the URL
- "location:"
- "[id]:"
- "https://"
- octodex.gi
- thub.com/
- images/
- dojocat.jp
- "g  \"The"
- "Dojocat\""
- "```"
- "Here's our"
- logo (
- hover to
- see the
- title text
- "):"
- Inline-
- "style:"
- "!["
- "alt text]("
- "https://"
- github.com
- /adam-p/
- markdown-
- here/raw/
- master/src
- /common/
- images/
- icon48.png
- "\"Logo"
- Title Text
- "1\")"
- Reference-
- "style:"
- "!["
- "alt text]["
- "logo]"
- "[logo]:"
- "https://"
- github.com
- /adam-p/
- markdown-
- here/raw/
- master/src
- /common/
- images/
- icon48.png
- "\"Logo"
- Title Text
- "2\""
- "!["
- "Minion]("
- "https://"
- octodex.gi
- thub.com/
- images/
- minion.png
- )
- "!["
- Stormtroop
- "ocat]("
- "https://"
- octodex.gi
- thub.com/
- images/
- stormtroop
- "ocat.jpg \""
- The
- Stormtroop
- "ocat\")"
- Like links
- ", Images"
- also have
- a footnote
- style
- syntax
- "!["
- "Alt text]["
- "id]"
- With a
- reference
- later in
- the
- document
- defining
- the URL
- "location:"
- "[id]:"
- "https://"
- octodex.gi
- thub.com/
- images/
- dojocat.jp
- "g  \"The"
- "Dojocat\""
- "------"
- "# ["
- "Footnotes]"
- "(https://"
- github.com
- /markdown-
- it/
- markdown-
- it-
- footnote)
- "```"
- Footnote 1
- "link[^"
- "first]."
- Footnote 2
- "link[^"
- "second]."
- Inline
- "footnote^["
- Text of
- inline
- "footnote]"
- definition
- "."
- Duplicated
- footnote
- "reference["
- "^second]."
- "[^first]:"
- Footnote *
- "*can have"
- markup**
- and
- multiple
- paragraphs
- "."
- "[^second]:"
- Footnote
- "text.\n```"
- Footnote 1
- "link[^"
- "first]."
- Footnote 2
- "link[^"
- "second]."
- Inline
- "footnote^["
- Text of
- inline
- "footnote]"
- definition
- "."
- Duplicated
- footnote
- "reference["
- "^second]."
- "[^first]:"
- Footnote *
- "*can have"
- markup**
- and
- multiple
- paragraphs
- "."
- "[^second]:"
- Footnote
- text.
- "------"
- "# Code and"
- Syntax
- Highlighti
- ng
- "```"
- "Inline `"
- "code` has"
- "`back-"
- ticks
- "around` it"
- ".\n```"
- "Inline `"
- "code` has"
- "`back-"
- ticks
- "around` it"
- "."
- "```c#"
- using
- System.IO.
- Compressio
- n;
- "#pragma"
- warning
- disable
- "414, 3021"
- namespace
- MyApplicat
- "ion\n{"
- "[Obsolete("
- "\"...\")]"
- class
- "Program :"
- IInterface
- "{"
- public
- static
- List<int>
- JustDoIt(
- int count)
- "{"
- Console.Wr
- "iteLine($\""
- "Hello {"
- "Name}!\");"
- return new
- List<int>(
- "new int[]"
- "{ 1, 2, 3"
- "})"
- "}\n    }\n}"
- "```"
- "```css"
- "@font-face"
- "{"
- font-
- "family:"
- Chunkfive;
- "src: url('"
- Chunkfive.
- "otf');\n}"
- "body, ."
- "usertext {"
- "color: #"
- F0F0F0;
- background
- ": #600;"
- font-
- "family:"
- "Chunkfive,"
- "sans;\n}"
- "@import"
- url(
- print.css)
- ;
- "@media"
- "print {"
- "a[href^="
- "http]::"
- "after {"
- "content:"
- attr(href)
- "}\n}\n```"
- "```"
- javascript
- function $
- initHighli
- "ght(block,"
- "cls) {"
- "try {"
- if (
- cls.search
- "(/\\bno\\-"
- "highlight\\"
- b/) != -1)
- return
- process(
- "block,"
- "true, 0x0F"
- ) +
- "` class=\"$"
- "{cls}\"`;"
- "} catch (e"
- ") {"
- /* handle
- exception
- "*/\n  }"
- for (var i
- "= 0 / 2; i"
- "<"
- classes.le
- ngth; i++)
- "{"
- if (
- checkCondi
- tion(
- "classes[i]"
- ) ===
- undefined)
- console.lo
- "g('"
- "undefined'"
- ");\n  }\n}"
- export  $
- initHighli
- "ght;\n```"
- "```php"
- require_on
- "ce 'Zend/"
- Uri/
- "Http.php';"
- namespace
- "Location\\"
- Web;
- interface
- "Factory\n{"
- static
- function
- _factory()
- ";\n}"
- abstract
- class URI
- extends
- BaseURI
- implements
- "Factory\n{"
- abstract
- function
- test();
- public
- static $
- st1 = 1;
- const ME =
- "\"Yo\";"
- var $list
- "= NULL;"
- private $
- var;
- /**
- "* Returns"
- a URI
- "*"
- "* @return"
- URI
- "*/"
- static
- public
- function
- _factory($
- stats =
- "array(), $"
- "uri = '"
- "http')"
- "{"
- echo
- __METHOD__
- ;
- $uri =
- "explode(':"
- "', $uri,"
- 0b10);
- $
- schemeSpec
- ific =
- isset($uri
- "[1]) ?"
- "$uri[1] :"
- "'';"
- "$desc = '"
- Multi
- line
- descriptio
- "n';"
- //
- Security
- check
- if (!
- ctype_alnu
- m($scheme)
- ") {"
- throw new
- Zend_Uri_E
- "xception('"
- Illegal
- "scheme');"
- "}"
- $this->var
- "= 0 - self"
- "::$st;"
- $this->
- list =
- list(Array
- "(\"1\"=> 2,"
- "2=>self::"
- "ME, 3 => \\"
- "Location\\"
- "Web\\URI::"
- class));
- "return ["
- "'uri'   =>"
- "$uri,"
- "'value' =>"
- "null,"
- "];\n    }\n}"
- "echo URI::"
- ME .
- "URI::$st1;"
- __halt_com
- piler () ;
- datahere
- datahere
- datahere *
- "/\ndatahere"
- "```"
- "------"
- "# Tables"
- "```"
- Colons can
- be used to
- align
- columns.
- "| Tables"
- "| Are"
- "| Cool  |"
- "| --------"
- "----- |:--"
- "----------"
- "-:| -----:"
- "|"
- "| col 3 is"
- "| right-"
- aligned |
- $1600 |
- "| col 2 is"
- "| centered"
- "|   $12 |"
- "| zebra"
- stripes |
- are neat
- "|    $1 |"
- There must
- be at
- least 3
- dashes
- separating
- each
- header
- cell.
- The outer
- pipes (|)
- are
- "optional,"
- and you
- "don't need"
- to make
- the
- raw
- Markdown
- line up
- prettily.
- You can
- also use
- inline
- Markdown.
- Markdown |
- Less |
- Pretty
- "--- | ---"
- "| ---"
- "*Still* |"
- "`renders`"
- "| **nicely"
- "**"
- 1 | 2 | 3
- "| First"
- Header  |
- Second
- Header |
- "| --------"
- "----- | --"
- "----------"
- "- |"
- "| Content"
- Cell  |
- Content
- Cell  |
- "| Content"
- Cell  |
- Content
- Cell  |
- "| Command"
- "|"
- Descriptio
- n |
- "| --- | --"
- "- |"
- "| git"
- status |
- List all
- new or
- modified
- files |
- "| git diff"
- "| Show"
- file
- difference
- s that
- "haven't"
- been
- staged |
- "| Command"
- "|"
- Descriptio
- n |
- "| --- | --"
- "- |"
- "| `git"
- "status` |"
- List all *
- new or
- modified*
- files |
- "| `git"
- "diff` |"
- Show file
- difference
- s that **
- "haven't"
- been**
- staged |
- "| Left-"
- aligned |
- Center-
- aligned |
- Right-
- aligned |
- "| :---"
- "|     :---"
- ":      |"
- "---: |"
- "| git"
- status   |
- git status
- "| git"
- status
- "|"
- "| git diff"
- "| git diff"
- "| git diff"
- "|"
- "| Name"
- "|"
- Character
- "|"
- "| ---"
- "| ---"
- "|"
- "| Backtick"
- "| `"
- "|"
- "| Pipe"
- "| \\|"
- "|\n```"
- Colons can
- be used to
- align
- columns.
- "| Tables"
- "| Are"
- "| Cool"
- "|"
- "| --------"
- "----- |:--"
- "----------"
- "-:| -----:"
- "|"
- "| col 3 is"
- "| right-"
- aligned |
- $1600 |
- "| col 2 is"
- "| centered"
- "|   $12 |"
- "| zebra"
- stripes |
- are neat
- "|    $1 |"
- There must
- be at
- least 3
- dashes
- separating
- each
- header
- cell.
- The outer
- pipes (|)
- are
- "optional,"
- and you
- "don't need"
- to make
- the
- raw
- Markdown
- line up
- prettily.
- You can
- also use
- inline
- Markdown.
- Markdown |
- Less |
- Pretty
- "--- | ---"
- "| ---"
- "*Still* |"
- "`renders`"
- "| **nicely"
- "**"
- 1 | 2 | 3
- "| First"
- Header  |
- Second
- Header |
- "| --------"
- "----- | --"
- "----------"
- "- |"
- "| Content"
- Cell  |
- Content
- Cell  |
- "| Content"
- Cell  |
- Content
- Cell  |
- "| Command"
- "|"
- Descriptio
- n |
- "| --- | --"
- "- |"
- "| git"
- status |
- List all
- new or
- modified
- files |
- "| git diff"
- "| Show"
- file
- difference
- s that
- "haven't"
- been
- staged |
- "| Command"
- "|"
- Descriptio
- n |
- "| --- | --"
- "- |"
- "| `git"
- "status` |"
- List all *
- new or
- modified*
- files |
- "| `git"
- "diff` |"
- Show file
- difference
- s that **
- "haven't"
- been**
- staged |
- "| Left-"
- aligned |
- Center-
- aligned |
- Right-
- aligned |
- "| :---"
- "|     :---"
- ":      |"
- "---: |"
- "| git"
- status   |
- git status
- "| git"
- status
- "|"
- "| git diff"
- "| git diff"
- "| git diff"
- "|"
- "| Name"
- "|"
- Character
- "|"
- "| ---"
- "| ---"
- "|"
- "| Backtick"
- "| `"
- "|"
- "| Pipe"
- "| \\|"
- "|\n\n------"
- "#"
- Blockquote
- s
- "```"
- ">"
- Blockquote
- s are very
- handy in
- email to
- emulate
- reply text
- "."
- "> This"
- line is
- part of
- the same
- quote.
- Quote
- break.
- "> This is"
- a very
- long line
- that will
- still be
- quoted
- properly
- when it
- wraps.
- Oh boy
- "let's keep"
- writing to
- make sure
- this is
- long
- enough to
- actually
- wrap for
- everyone.
- "Oh, you"
- can *put*
- "**Markdown"
- "** into a"
- blockquote
- "."
- ">"
- Blockquote
- s can also
- be nested.
- ".."
- ">> ...by"
- using
- additional
- greater-
- than signs
- right next
- to each
- other...
- "> > > ..."
- or with
- spaces
- between
- arrows.
- "```"
- ">"
- Blockquote
- s are very
- handy in
- email to
- emulate
- reply text
- ".\n>"
- This line
- is part of
- the same
- quote.
- Quote
- break.
- "> This is"
- a very
- long line
- that will
- still be
- quoted
- properly
- when it
- wraps.
- Oh boy
- "let's keep"
- writing to
- make sure
- this is
- long
- enough to
- actually
- wrap for
- everyone.
- "Oh, you"
- can *put*
- "**Markdown"
- "** into a"
- blockquote
- "."
- ">"
- Blockquote
- s can also
- be nested.
- "..\n>"
- "> ...by"
- using
- additional
- greater-
- than signs
- right next
- to each
- other...
- "> >"
- "> ...or"
- with
- spaces
- between
- arrows.
- "------"
- "# Inline"
- HTML
- "```\n<dl>"
- "<dt>"
- Definition
- list</dt>
- "<dd>Is"
- something
- people use
- sometimes.
- "</dd>"
- "<dt>"
- Markdown
- in HTML</
- dt>
- "<dd>Does *"
- not* work
- "**very**"
- well.
- Use HTML <
- em>tags</
- em>.</dd>
- "</dl>\n```"
- "<dl>"
- "<dt>"
- Definition
- list</dt>
- "<dd>Is"
- something
- people use
- sometimes.
- "</dd>"
- "<dt>"
- Markdown
- in HTML</
- dt>
- "<dd>Does *"
- not* work
- "**very**"
- well.
- Use HTML <
- em>tags</
- em>.</dd>
- "</dl>"
- "------"
- "#"
- Horizontal
- Rules
- "```"
- Three or
- more...
- "---"
- Hyphens
- "***"
- Asterisks
- ___
- Underscore
- "s\n```"
- Three or
- more...
- "---"
- Hyphens
- "***"
- Asterisks
- ___
- Underscore
- "s\n\n------"
- "# YouTube"
- Videos
- "```"
- "<a href=\""
- "http://"
- www.youtub
- e.com/
- watch?
- feature=
- player_emb
- edded&v=
- YOUTUBE_VI
- DEO_ID_HER
- "E\" target="
- "\"_blank\">"
- "<img src=\""
- "http://"
- img.youtub
- e.com/vi/
- YOUTUBE_VI
- DEO_ID_HER
- "E/0.jpg\""
- "alt=\"IMAGE"
- ALT TEXT
- "HERE\""
- "width=\"240"
- "\" height=\""
- "180\""
- "border=\"10"
- "\">\n</a>"
- "```"
- "<a href=\""
- "http://"
- www.youtub
- e.com/
- watch?
- feature=
- player_emb
- edded&v=
- YOUTUBE_VI
- DEO_ID_HER
- "E\" target="
- "\"_blank\">"
- "<img src=\""
- "http://"
- img.youtub
- e.com/vi/
- YOUTUBE_VI
- DEO_ID_HER
- "E/0.jpg\""
- "alt=\"IMAGE"
- ALT TEXT
- "HERE\""
- "width=\"240"
- "\" height=\""
- "180\""
- "border=\"10"
- "\">\n</a>"
- "```\n[!["
- IMAGE ALT
- "TEXT HERE]"
- "(http://"
- img.youtub
- e.com/vi/
- YOUTUBE_VI
- DEO_ID_HER
- "E/0.jpg)]("
- "http://"
- www.youtub
- e.com/
- watch?
- v=
- YOUTUBE_VI
- DEO_ID_HER
- "E)\n```"
- "[!["
- IMAGE ALT
- "TEXT HERE]"
- "(https://"
- upload.wik
- imedia.org
- /wikipedia
- /commons/
- thumb/e/ef
- /
- YouTube_lo
- go_2015.
- svg/1200px
- "-"
- YouTube_lo
- go_2015.
- "svg.png)]("
- "https://"
- www.youtub
- e.com/
- watch?
- v=
- ciawICBvQo
- E)

